packages/datadog-instrumentations/src/sharedb.js

```diff
@@ -0,0 +1,78 @@
+'use strict'
+
+const {
+  channel,
+  addHook,
+  AsyncResource
+} = require('./helpers/instrument')
+const shimmer = require('../../datadog-shimmer')
+
+/**
+ * @description The enum values in this map are not exposed from ShareDB, so the keys are hard-coded here.
+ * The values were derived from: https://github.com/share/sharedb/blob/master/lib/client/connection.js#L196
+ */
+const READABLE_ACTION_NAMES = {
+  hs: 'handshake',
+  qf: 'query-fetch',
+  qs: 'query-subscribe',
+  qu: 'query-unsubscribe',
+  bf: 'bulk-fetch',
+  bs: 'bulk-subscribe',
+  bu: 'bulk-unsubscribe',
+  f: 'fetch',
+  s: 'subscribe',
+  u: 'unsubscribe',
+  op: 'op',
+  nf: 'snapshot-fetch',
+  nt: 'snapshot-fetch-by-ts',
+  p: 'presence-broadcast',
+  pr: 'presence-request',
+  ps: 'presence-subscribe',
+  pu: 'presence-unsubscribe'
+}
+
+addHook({ name: 'sharedb', versions: ['>=1'], file: 'lib/agent.js' }, Agent => {
+  const startCh = channel('apm:sharedb:_handleMessage:start')
+  const asyncEndCh = channel('apm:sharedb:_handleMessage:async-end')
+  const endCh = channel('apm:sharedb:_handleMessage:end')
+  const errorCh = channel('apm:sharedb:_handleMessage:error')
+
+  shimmer.wrap(Agent.prototype, '_handleMessage', origHandleMessageFn => function (request, callback) {
+    const asyncResource = new AsyncResource('bound-anonymous-fn')
+    const action = request.a
+
+    const actionName = getReadableActionName(action)
+
+    startCh.publish({ actionName, request })
+
+    callback = asyncResource.bind(callback)
```

Comment:
Do you actually need this, given that you're wrapping everything in a bind on line 50 anyway?

Reply:
Nevermind, this was explained to me.